ports/litex/common-hal/microcontroller/Processor.c

```diff
@@ -0,0 +1,54 @@
+/*
+ * This file is part of the MicroPython project, http://micropython.org/
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2017 Dan Halbert for Adafruit Industries
+ * Copyright (c) 2019 Lucian Copeland for Adafruit Industries
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include <math.h>
+#include "common-hal/microcontroller/Processor.h"
+#include "py/runtime.h"
+#include "supervisor/shared/translate.h"
+
+#include "csr.h"
+#include "generated/soc.h"
+
+#define STM32_UUID ((uint32_t *)0xf05bf05b)
+
+float common_hal_mcu_processor_get_temperature(void) {
+    return NAN;
+}
+
+float common_hal_mcu_processor_get_voltage(void) {
+    return NAN;
+}
+
+uint32_t common_hal_mcu_processor_get_frequency(void) {
+    return CONFIG_CLOCK_FREQUENCY;
+}
+
+void common_hal_mcu_processor_get_uid(uint8_t raw_id[]) {
+    for (int i=0; i<3; i++) {
+        ((uint32_t*) raw_id)[i] = STM32_UUID[i];
```

Comment:
I don't think this will work because it looks like the STM32 code.

Reply:
yeah this is an STM32 ID at a ST specific address, it won't do anything here.

Reply:
I've updated it to generate a "serial number" based on the contents of the VERSION field.

Reply:
Is this unique for every ICE40 or the same for all? Ideally it'd be different since this is what we use as the USB serial number.

Reply:
There's nothing that's unique for every ICE40.  And we don't burn in serial numbers as part of Fomu.

Reply:
What happens if you plug two in at the same time? I bet the flash part has a unique id you could use. I think that's pretty standard.

Reply:
We don't program a serial number into the flash.

Only the W25Q128 has a serial number, and that was only used on EVT boards.  PVT and hacker boards use the AT25SF161, MX25R1635F, or GD25Q16C, none of which have a serial number programmed from the manufacturer.

If you plug two in at the same time, they will both have the same serial number.